interfaces/builtin/mount_control.go

```diff
@@ -368,22 +416,34 @@ func validateMountOptions(mountInfo *MountInfo) error {
 	if len(mountInfo.options) == 0 {
 		return errors.New(`mount-control "options" cannot be empty`)
 	}
-	// With "functionfs" none of the valid "options" can be harmful so no need to check against allowedMountOptions
-	if mountInfo.isType("functionfs") {
-		if err := validateNoAppArmorRegexpWithError(`cannot use mount-control "option" attribute`, mountInfo.options...); err != nil {
-			return err
+	if err := validateNoAppArmorRegexpWithError(`cannot use mount-control "option" attribute`, mountInfo.options...); err != nil {
+		return err
+	}
+	var types []string
+	if mountInfo.hasType() {
+		fsExclusiveOption := optionIncompatibleWithFsType(mountInfo.options)
+		if fsExclusiveOption != "" {
+			return fmt.Errorf(`mount-control option %q is incompatible with specifying filesystem type`, fsExclusiveOption)
 		}
+		types = mountInfo.types
 	} else {
-		for _, o := range mountInfo.options {
-			if !strutil.ListContains(allowedMountOptions, o) {
+		types = defaultFSTypes
+	}
+	for _, o := range mountInfo.options {
+		if strutil.ListContains(allowedKernelMountOptions, o) {
+			continue
+		}
+		if strutil.ListContains(allowedUserspaceMountOptions, o) {
+			continue
+		}
+		optionName := strings.Split(o, "=")[0]
```

Comment:
Update: with commit b1d4a36 I've switched to using `SplitAfter` rather than `Split`, so it's no longer possible to validate `sync=foo` as the `sync` kernel option (for example). This makes me feel better about stripping the optarg before validating kernel mount options, but I think the question about how apparmor would handle kernel mount options with optargs is still unknown.

On the one hand, leaving things as they are now enforces that kernel options should always be validated in their entirety (which is true now and maybe forever, since kernel mount options are always converted to flags --- actually this is a really good reason to treat kernel options as fundamentally different from the others and not strip optargs until after kernel option validation). On the other hand, it would be more "consistent" to strip optargs before validation, but that would mean that kernel option optargs should probably also be stripped when forming apparmor rules, about which the behavior is not well-defined.

Because kernel options never have optargs and thus there's no definition for how apparmor should deal with them, I'm going to leave the optarg stripping where it is now (after kernel option validation).

Reply:
Thanks for this very thoughtful comment - I think your reasoning makes sense and I agree with the approach.